---
title: "ELMER.data: Supporting data for the ELMER package"
author: "Lijing Yao, Ben Berman [aut], Peggy Farnham [aut]Hui Shen [ctb], Peter Laird [ctb], Simon Coetzee [ctb], Tiago Chedraoui Silva [ctb]"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    number_sections: true
    toc_depth: 3
    highlight: haddock
    fig_caption: yes
  BiocStyle::pdf_document2:
    toc: true
    number_sections: true
    toc_depth: 3
    highlight: haddock
    fig_caption: yes     
    
references:
- id: ref1
  title: HOCOMOCO a comprehensive collection of human transcription factor binding sites models
  author: 
  - family: Kulakovskiy, Ivan V and Medvedeva, Yulia A and Schaefer, Ulf and Kasianov, Artem S and Vorontsov, Ilya E and Bajic, Vladimir B and Makeev, Vsevolod
    given:
  journal: Nucleic acids research
  volume: 41
  number: D1
  pages: D195--D202
  issued:
    year: 2013    

- id: heinz2010simple
  title: Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities
  author: 
  - family: Heinz, Sven and Benner, Christopher and Spann, Nathanael and Bertolino, Eric and Lin, Yin C and Laslo, Peter and Cheng, Jason X and Murre, Cornelis and Singh, Harinder and Glass, Christopher K
    given:
  journal: Molecular cell
  volume: 38
  number: 4
  pages: 576--589
  issued:
    year: 2010    
    
- id: ELMER
  title: Inferring regulatory element landscapes and transcription factor networks from cancer methylomes
  author: 
  - family: Yao, L., Shen, H., Laird, P. W., Farnham, P. J., & Berman, B. P. 
    given:
  journal: Genome biology
  volume: 16
  number: 1
  pages: 105
  issued:
    year: 2015 

- id: wingender2014tfclass
  title: TFClass a classification of human transcription factors and their rodent orthologs
  author: 
  - family: Wingender, E., Schoeps, T., Haubrock, M., & DÃ¶nitz, J
    given:
  journal: Nucleic acids research
  pages: gku1064
  issued:
    year: 2014 

- id: zhou2016comprehensive
  title: Comprehensive characterization, annotation and innovative use of Infinium DNA methylation BeadChip probes
  author: 
  - family: Zhou, Wanding and Laird, Peter W and Shen, Hui
    given:
  journal: Nucleic Acids Research
  pages: gkw967
  issued:
    year: 2016


vignette: >
  %\VignetteIndexEntry{ELMER.data: Supporting data for the ELMER package}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---   

```{r, echo = FALSE,hide=TRUE, message=FALSE,warning=FALSE}
devtools::load_all(".")
```

# Introduction
This document provides an introduction of the \Biocpkg{ELMER.data}, which contains 
supporting data for \Biocpkg{ELMER} [@ELMER]. \Biocpkg{ELMER} is package using DNA methylation to 
identify enhancers, and correlates enhancer state with expression of nearby genes 
to identify one or more transcriptional targets. Transcription factor (TF) binding 
site analysis of enhancers is coupled with expression analysis of all TFs to 
infer upstream regulators. \Biocpkg{ELMER.data} provide 2 necessary data for 
\Biocpkg{ELMER} analysis:

1. Probes.motif: motif occurences within -/+250bp of probe sites on HM450K/EPIC array aligned against hg19/hg38.
2. Union.enhancer: A comprehensive list of genomic strong enhancers.

## Installing and loading ELMER.data
To install this package, start R and enter

```{r, eval = FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("ELMER.data")
library("ELMER.data")
library("GenomicRanges")
```

# Contents

## Probes information

Probes information were retrieved from (http://zwdzwd.github.io/InfiniumAnnotation)[http://zwdzwd.github.io/InfiniumAnnotation] [@zhou2016comprehensive]
```{r, eval=FALSE, include=TRUE}
for(plat in c("450K","EPIC")) {
  for(genome in c("hg38","hg19")) {
    base <- "http://zwdzwd.io/InfiniumAnnotation/current/"
    if(plat == "EPIC") {
      annotation <- paste0(base,"EPIC/EPIC.manifest.rda")
    } else {
      annotation <- paste0(base,"hm450/hm450.manifest.rda")
    }
    if(genome == "hg38") annotation <- gsub(".rda",".hg38.rda", annotation)
    if(!file.exists(basename(annotation))) {
      if(Sys.info()["sysname"] == "Windows") mode <- "wb" else  mode <- "w"
      downloader::download(annotation, basename(annotation), mode = mode)
    }
  }
}
```
```{r}
data("EPIC.manifest")
data("EPIC.manifest.hg38")
data("hm450.manifest.hg38")
data("hm450.manifest")
```

## Probes.motif

Probes.motif provides information for motif occurences within -/+250bp of probe 
sites on HM450K/EPIC array. HOMER [@heinz2010simple] was used with a p-value < 0.0001 to scan a +/- 250bp 
region around each probe on HM450K/EPIC using HOCOMOCO V10 motif position weight 
matrices (PWMs) which provides transcription factor (TF) binding models for more than 600 human TFs.
This data set is used in get.enriched.motif function in \Biocpkg{ELMER} to calculate 
Odds Ratio of motif enrichments for a given set of probes. This data is storaged 
in a sparse matrix with wuth 640 columns, there is one matrix for HM450K aligned to hg19, one for HM450K  aligned to hg38,
one for EPIC aligned to hg19, one for EPIC  aligned to hg38. Each row is each probe regions (annotation of the regions used can be found in 
[this repository](http://zwdzwd.github.io/InfiniumAnnotation)) and each 
column is motif from [http://hocomoco.autosome.ru/](HOCOMOCO) [@ref1]. The value 1 indicates the occurrence 
of a motif in a particular probe and 0 means no occurrence.

```{r}
data("Probes.motif.hg19.450K")
dim(Probes.motif.hg19.450K)
str(Probes.motif.hg19.450K)
```

```{r}
data("Probes.motif.hg38.450K")
dim(Probes.motif.hg38.450K)
str(Probes.motif.hg38.450K)
```

```{r}
data("Probes.motif.hg19.EPIC")
dim(Probes.motif.hg19.EPIC)
str(Probes.motif.hg19.EPIC)
```

```{r}
data("Probes.motif.hg38.EPIC")
dim(Probes.motif.hg38.EPIC)
str(Probes.motif.hg38.EPIC)
```

The following code was used to create the objects:
```{r, eval=FALSE, include=TRUE}
getInfiniumAnnotation <- function(plat = "450K", genome = "hg38"){
    message("Loading object: ",file)
    newenv <- new.env()
    if(plat == "EPIC" & genome == "hg19") data("EPIC.manifest", package = "ELMER.data",envir=newenv)
    if(plat == "EPIC" & genome == "hg38") data("EPIC.manifest.hg38", package = "ELMER.data",envir=newenv)
    if(plat == "450K" & genome == "hg19") data("hm450.manifest", package = "ELMER.data",envir=newenv)
    if(plat == "450K" & genome == "hg38") data("hm450.manifest.hg38", package = "ELMER.data",envir=newenv)
    annotation <- get(ls(newenv)[1],envir=newenv)   
    return(annotation)  
}
# To find for each probe the know motif we will use HOMER software (http://homer.salk.edu/homer/)
# Step:
# 1 - get DNA methylation probes annotation with the regions
# 2 - Make a bed file from it
# 3 - Execute section: Finding Instance of Specific Motifs from http://homer.salk.edu/homer/ngs/peakMotifs.html to the HOCOMOCO TF motifs
# Also, As HOMER is using more RAM than the available we will split the files in to 100k probes.
# Obs: for each probe we create a winddow of 500 bp (-size 500) around it. This might lead to false positives, but will not have false negatives.
# The false posives will be removed latter with some statistical tests.
TFBS.motif <- "http://hocomoco.autosome.ru/final_bundle/HUMAN/mono/HOCOMOCOv10_HUMAN_mono_homer_format_0.0001.motif"
if(!file.exists(basename(TFBS.motif))) downloader::download(TFBS.motif,basename(TFBS.motif))
for(plat in c("450K","EPIC")){
    for(gen in c("hg19","hg38")){
      
      file <- paste0(plat,gen,".txt")
      print(file)
      if(!file.exists(file)){
        # STEP 1
        gr <- getInfiniumAnnotation(plat = plat,genome =  gen)
        
        # This will remove masked probes. They have poor quality and might be arbitrarily positioned (Wanding Zhou)
        print(table(gr$MASK.general))
        gr <- gr[!gr$MASK.general]
        print(table(gr$MASK.general))
        
        df <- data.frame(seqnames=seqnames(gr),
                         starts=as.integer(start(gr)),
                         ends=end(gr),
                         names=names(gr),
                         scores=c(rep(".", length(gr))),
                         strands=strand(gr))
        step <- 50000 # nb of lines in each file 50K was selected to not explode RAM
        n <- nrow(df)
        for(j in 0:floor(n/step)){
          # STEP 2
          file.aux <- paste0(plat,gen,"_",j,".bed")
          if(!file.exists(gsub(".bed",".txt",file.aux))){
            end <- ifelse(((j + 1) * step) > n, n,((j + 1) * step))
            write.table(df[((j * step) + 1):end,], file = file.aux, col.names = F, quote = F,row.names = F,sep = "\t")
            
            # STEP 3
            cmd <- paste("source ~/.bash_rc; annotatePeaks.pl" ,file.aux, gen, "-m", basename(TFBS.motif), "-size 500 -cpu 12 >", gsub(".bed",".txt",file.aux))
            system(cmd)
          }
        }
        # We will merge the results from each file into one
        peaks <- NULL
        for(j in 0:floor(n/step)){
          aux <-  readr::read_tsv(paste0(plat,gen,"_",j,".txt"))
          colnames(aux)[1] <- "PeakID"
          if(is.null(peaks)) {
            peaks <- aux
          } else {
            peaks <- rbind(peaks, aux)
          }
        }
        readr::write_tsv(peaks,path=file,col_names = TRUE)
        print("DONE!")
        gc()
      }
  }
}

# This code will read the table with the motifs, save it as a sparce matrix
# and save all as a .rda that will be placed in ELMER.data
for(plat in c("450K","EPIC")){
    for(gen in c("hg19","hg38")){
        file <- paste0(plat,gen,".txt")
        motifs <- readr::read_tsv(file)
        # From 1 to 21 we have annotations then we have 640 motifs
        matrix <- Matrix::Matrix(0, nrow = nrow(motifs), ncol = 640,sparse = TRUE)
        colnames(matrix) <- gsub(" Distance From Peak\\(sequence,strand,conservation\\)","",colnames(motifs)[-c(1:21)])
        rownames(matrix) <- motifs$PeakID
        matrix[!is.na(motifs[,-c(1:21)])] <- 1
        matrix <- as(matrix, "nsparseMatrix")
        assign(paste0("Probes.motif.",gen,".",plat),matrix) # For each probe that there is a bind to the motif, we will add as 1 in the matrix. (Are hg38, hg19,450k,epic matrices equal?)
        rm(matrix)
        rm(motifs)
        gc()
    }
}
save(Probes.motif.hg19.450K, file = "Probes.motif.hg19.450K.rda", compress = "xz")
save(Probes.motif.hg38.450K, file = "Probes.motif.hg38.450K.rda", compress = "xz")
save(Probes.motif.hg19.EPIC, file = "Probes.motif.hg19.EPIC.rda", compress = "xz")
save(Probes.motif.hg38.EPIC, file = "Probes.motif.hg38.EPIC.rda", compress = "xz")
```

## motif.relevant.TFs

motif.relavent.TFs list each TF motif and TFs that binding the motis. Multiple TFs may
recognize a same motif such as TF family.  
The association between each motif famil and transcription factor was created using the 
(HOCOMOCO)[http://hocomoco.autosome.ru/human/mono] which TF structural families 
was created according to TFClass [@wingender2014tfclass]
This data is stored as a list whose elements 
are motifs and contents for each element are TFs which recognize the same motif that
is the name of the element. This data is used in function get.TFs in \Biocpkg{ELMER} 
to identify the real regulator TF whose motif is enriched in a given set of probes 
and expression associate with average DNA methylation of these motif sites.

The following code was used to create the objects:
```{r, eval=FALSE, include=TRUE}
library(plyr)
library(rvest)
library(xml2)
createMotifRelevantTfs <- function(){
  if(!file.exists("motif.relavent.TFs.rda")){
    # Download from http://hocomoco.autosome.ru/human/mono
    tf.family <- "http://hocomoco.autosome.ru/human/mono" %>% read_html()  %>%  html_table()
    tf.family <- tf.family[[1]]
    # Split TF for each family, this will help us map for each motif which are the some ones in the family
    # basicaly: for a TF get its family then get all TF in that family
    family <- split(tf.family,f = tf.family$`TF family`)
    motif.relavent.TFs <- plyr::alply(tf.family,1, function(x){  
      f <- x$`TF family`
      if(f == "") return(x$`Transcription factor`) # Case without family, we will get only the same object
      return(unique(family[as.character(f)][[1]]$`Transcription factor`))
    },.progress = "text")
    #names(motif.relavent.TFs) <- tf.family$`Transcription factor`
    names(motif.relavent.TFs) <- tf.family$Model
    # Cleaning object
    attr(motif.relavent.TFs,which="split_type") <- NULL
    attr(motif.relavent.TFs,which="split_labels") <- NULL
    save(motif.relavent.TFs, file = "motif.relevant.TFs.rda", compress = "xz")
  }
}
```


```{r}
data("motif.relevant.TFs")
motif.relavent.TFs[1:2]
```

## Union.enhancer

Union.enhancer is a comprehensive list of genomic strong enhancers which comes 
from a combination of enhancers from the Roadmap Epigenomics Mapping Consortium (REMC) 
and the Encyclopedia of DNA Elements (ENCODE) Project, in which enhancers were 
identified using ChromHMM for 98 tissues or cell lines. We used the union of 
genomic elements labeled as EnhG1, EnhG2, EnhA1 or EnhA2 (representing intergenic 
and intragenic active enhancers) in any of the 98 cell types, resulting in a total 
of 389,967 non-overlapping enhancer regions. FANTOM5 enhancers (43,011) identified 
by eRNAs for 400 distinct cell types were added to this list. This data is stored 
as a GRanges object contains coordinates of this human comprehensive genomic enhancer set.
It will be used in get.feature.probe in \Biocpkg{ELMER} to identify the distal enhancer
probes which are at least 2Kb awary from TSS regions and overlap with comprehensive 
enhancers.

```{r}
data("Union.enhancer.hg19")
Union.enhancer.hg19
```

The data used to create the Union.enhancer object was aligned 
against the reference genome hg19. 
In order to align it against hg38 we used liftOver.

```{r, eval=FALSE, include=TRUE}
data("Union.enhancer.hg19")
file <- "http://hgdownload.cse.ucsc.edu/gbdb/hg19/liftOver/hg19ToHg38.over.chain.gz"
downloader::download(file,basename(file))
R.utils::gunzip(basename(file))
chain <- import.chain("hg19ToHg38.over.chain")
Union.enhancer.hg38 <- unlist(rtracklayer::liftOver(Union.enhancer.hg19, chain))
Union.enhancer.hg38
```
```{r}
data("Union.enhancer.hg38")
Union.enhancer.hg38
```

# Session Information
******
```{r sessionInfo}
sessionInfo()
```

# References
